internal/e2e/testcase.go

```diff
@@ -0,0 +1,314 @@
+		t.Fatalf("Could not execute pipeline config template: tmpl.Execute(%q) error = %v", tmplInput, err)
+	}
+
+	pipeline, cleanupPipeline, err := createPipeline(ctx, client, name, pipelineCfg.String())
+	if err != nil {
+		t.Fatalf("Could not create pipeline with the following config in org %q: testHelper.createPipeline(%q, pipelineCfg) error = %v\n%s", targetOrg, name, err, pipelineCfg.String())
+	}
+	t.Cleanup(func() {
+		if err := cleanupPipeline(); err != nil {
+			t.Logf("Could not clean up pipeline %q (id = %s) in org %q: cleanup() = %v", pipeline.Slug, pipeline.ID, targetOrg, err)
+		}
+	})
+
+	return &testCase{
+		TB:             t,
+		fullName:       name,
+		bkClient:       client,
+		pipelineConfig: tmpl,
+		queue:          queue,
+		pipeline:       pipeline,
+	}
+}
+
+// triggerBuild creates a new build in the target pipeline. It returns the
+// build object. It also registers cleanups with t.Cleanup so that the build is
+// (usually) automatically cancelled if it is still running.
+// It calls t.Fatal if there was an error creating the build.
+func (tc *testCase) triggerBuild() *buildkite.Build {
+	tc.Helper()
+	ctx := tc.Context()
+
+	createBuild := buildkite.CreateBuild{
+		Author: buildkite.Author{
+			Email: authorEmail,
+			Name:  cmp.Or(authorName, "Agent E2E Tests"),
+		},
+		Commit:  "HEAD",
+		Branch:  "main",
+		Message: tc.fullName,
+	}
+
+	build, _, err := tc.bkClient.Builds.Create(ctx, targetOrg, tc.pipeline.Slug, createBuild)
+	if err != nil {
+		tc.Fatalf("tc.bkClient.Builds.Create(ctx, %q, %q, %v) error = %v", targetOrg, tc.pipeline.Slug, createBuild, err)
+	}
+
+	tc.Logf("Triggered a build at https://buildkite.com/%s/%s/builds/%d", targetOrg, tc.pipeline.Slug, build.Number)
+
+	tc.Cleanup(func() {
+		ctx := context.WithoutCancel(ctx) // allow cleanup after the test
+		_, err := tc.bkClient.Builds.Cancel(ctx, targetOrg, tc.pipeline.Slug, strconv.Itoa(build.Number))
+		if err != nil {
+			reasons := []string{
+				"already finished",
+				"already being canceled",
+				"already been canceled",
+				"No build found",
+			}
+			ignorable := slices.ContainsFunc(reasons, func(r string) bool {
+				return strings.Contains(err.Error(), r)
+			})
+			if ignorable {
+				return
+			}
+			tc.Logf("Couldn't cancel build %s: %v", build.ID, err)
+		}
+	})
+	return &build
+}
+
+// waitForBuild waits until the build is in a terminal state (passed, failed, canceled, etc). It polls the build once per second.
+func (tc *testCase) waitForBuild(ctx context.Context, build *buildkite.Build) (string, error) {
+	tick := time.Tick(time.Second)
+	for {
+		// The arg is called "id" but it needs the build number...
+		state, _, err := tc.bkClient.Builds.Get(ctx, targetOrg, tc.pipeline.Slug, strconv.Itoa(build.Number), &buildkite.BuildGetOptions{})
+		if err != nil {
+			return "", err
+		}
+		switch state.State {
+		case "passed", "failed", "canceled", "canceling":
+			return state.State, nil
+
+		case "scheduled", "running":
```

Comment:
Without looking into how bk client work, this is a bit confusing? Why is there a "scheduled" state on build? I thought it's a job thing 🤔 .